Task 4: HashTable

Iwirii opened the pull request:
Реализован ассоциативны массив, который использует pool-аллокатор для управления памятью.
Реализованы следующие функции:
- void hashtable_init(HashTable* table, size_t capacity, Allocator* allocator) — инициализация таблицы.
- void hashtable_insert(HashTable* table, const char* key, void* value) — вставка элемента.
- void* hashtable_get(HashTable* table, const char* key) — получение значения по ключу.
- void hashtable_del(HashTable* table, const char* key) — удаление по ключу
- void hashtable_free(HashTable* table) — освобождение памяти.

Фирстова Ксения Максимовна

Iwirii committed:
feat: add Makefile for hash table with linear allocator

Iwirii committed:
feat(hash-table): implement hash table with pool allocator

- Add complete hash table implementation with init/insert/get/delete/free operations
- Integrate with existing pool allocator without modifications
- Include comprehensive test suite covering basic ops and collisions
- Ensure memory safety with proper allocation checks

Iwirii committed:
Update main.yml

Iwirii committed:
style: fix indentation in Makefile code

Iwirii committed:
test: add allocator overflow and large volume tests

zaryanezrya commented on hash_table.c:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  entry->value = value;
+  entry->next = table->buckets[index];
```
Надо копировать (memcpy)
В тестах передаются значения по указателю на элемент. Элементы находятся на стеке. После завершения функции с переменными на стеке в листе остаются висячие указатели.

Iwirii committed:
fix(hash_table_test): prevent dangling pointers by copying values

Iwirii committed:
chore: apply code formatting with `make fmt`

Iwirii committed:
fix(hash_table): prevent dangling pointers by copying values

- Added value_size to hash_table_entry to store data size
- Modified hash_table_insert to copy values using memcpy
- Updated all deletion functions to free copied values
- Adjusted tests to pass value sizes
- Removed dependency on stack variables in tests

Fixes issue with dangling pointers when using stack variables

Iwirii committed:
chore: apply code formatting with make fmt

Iwirii committed:
Merge branch 'main' into HashTable
